### PR TITLE
FEM-2272 upgrade to Exo2.9.1

### PR DESCRIPTION
**Pull request by giladna**

### CODE REVIEW

File: playkit/src/main/java/com/kaltura/playkit/player/DummySurfaceWorkaroundTest.java [hunk 1]
```diff
@@ -117,7 +116,31 @@ private static void workaroundRequired(Context context, boolean b) {
 
     private static DefaultDrmSessionManager<FrameworkMediaCrypto> getDrmSessionManager(Handler mainHandler) {
         try {
-            return DefaultDrmSessionManager.newWidevineInstance(fakeDrmCallback, null, mainHandler, null);
+            DefaultDrmSessionManager defaultDrmSessionManager = DefaultDrmSessionManager.newWidevineInstance(fakeDrmCallback, null);
+            if (mainHandler != null) {
```
Comment (noamtamim) [inline playkit/src/main/java/com/kaltura/playkit/player/DummySurfaceWorkaroundTest.java]: Remove this section, don't need a listener.

File: playkit/src/main/java/com/kaltura/playkit/player/TrackSelectionHelper.java [hunk 2]
```diff
@@ -480,15 +478,19 @@ private SelectionOverride retrieveOverrideSelection(int[] uniqueId) {
...
-
                     for (int i = 0; i < audioTracks.size(); i++) {
 
                         audioTrack = audioTracks.get(i);
                         audioGroupIndex = getIndexFromUniqueId(audioTrack.getUniqueId(), GROUP_INDEX);
                         audioTrackIndex = getIndexFromUniqueId(audioTrack.getUniqueId(), TRACK_INDEX);
-
-                        if (audioGroupIndex == groupIndex && audioTrackIndex != TRACK_ADAPTIVE) {
-                            adaptiveTrackIndexesList.add(getIndexFromUniqueId(audioTrack.getUniqueId(), TRACK_INDEX));
+                        //TODO  - Validate this logic with more streams
```
Comment (noamtamim) [inline playkit/src/main/java/com/kaltura/playkit/player/TrackSelectionHelper.java]: Test with:
"http://cdnapi.kaltura.com/p/2215841/playManifest/entryId/1_w9zx2eti/format/mpegdash/protocol/http/a.mpd"
"http://cdnapi.kaltura.com/p/2215841/playManifest/entryId/1_w9zx2eti/format/applehttp/protocol/http/a.m3u8"

**Review by noamtamim — CHANGES_REQUESTED** [review]
Check offline+DRM.

File: playkit/src/main/java/com/kaltura/playkit/Player.java [hunk 3]
```diff
@@ -204,6 +204,13 @@
      */
     void replay();
 
+    /**
```
Comment (noamtamim) [inline playkit/src/main/java/com/kaltura/playkit/Player.java]: Remove retry.

File: playkit/src/main/java/com/kaltura/playkit/drm/MediaDrmSession.java [hunk 4]
```diff
@@ -63,7 +67,10 @@ void restoreKeys(byte[] keySetId) {
 
     FrameworkMediaDrm.KeyRequest getOfflineKeyRequest(byte[] initData, String mimeType) {
         try {
-            return mMediaDrm.getKeyRequest(mSessionId, initData, mimeType, MediaDrm.KEY_TYPE_OFFLINE, null);
+            DrmInitData.SchemeData schemeData = new DrmInitData.SchemeData(MediaSupport.WIDEVINE_UUID, mimeType, initData);
```
Comment (noamtamim) [inline playkit/src/main/java/com/kaltura/playkit/drm/MediaDrmSession.java]: Use Collections.singletonList().

**Review by noamtamim — CHANGES_REQUESTED** [review]

File: playkit/src/main/java/com/kaltura/playkit/Player.java (hunk 3 above)
Comment (giladna) [reply playkit/src/main/java/com/kaltura/playkit/Player.java]: done

File: playkit/src/main/java/com/kaltura/playkit/drm/MediaDrmSession.java (hunk 4 above)
Comment (giladna) [reply playkit/src/main/java/com/kaltura/playkit/drm/MediaDrmSession.java]: done

File: playkit/src/main/java/com/kaltura/playkit/player/TrackSelectionHelper.java (hunk 2 above)
Comment (giladna) [reply playkit/src/main/java/com/kaltura/playkit/player/TrackSelectionHelper.java]: looks fine with these streams

**Review by noamtamim — APPROVED** [review]

**Review by noamtamim — CHANGES_REQUESTED** [review]
Let's upgrade to 2.9.1 before merging, there are fixes that may be important.

File: playkit/src/main/java/com/kaltura/playkit/player/TrackSelectionHelper.java [hunk 5]
```diff
@@ -617,13 +617,12 @@ private boolean isAdaptive(int rendererIndex, int groupIndex) {
      * Validate and return parsed uniqueId.
      *
      * @param uniqueId - uniqueId to validate
-     * @return - parsed uniqueId in case of success.
-     * @throws IllegalArgumentException when uniqueId is illegal.
+     * @return IllegalArgumentException when uniqueId is illegal ele return null
      */
-    private int[] validateUniqueId(String uniqueId) throws IllegalArgumentException {
+    public IllegalArgumentException validateUniqueId(String uniqueId) {
```
Comment (noamtamim) [inline playkit/src/main/java/com/kaltura/playkit/player/TrackSelectionHelper.java]: Returning an exception from a function looks weird. Maybe instead, throw the message like before but be prepared to catch it in ExoPlayerWrapper.
So, no change in TrackSelectionHelper (make it like it was before this commit) and in ExoPlayerWrapper add `try..catch (IllegalArgumentException)` around changeTrack().

Comment (giladna) [reply playkit/src/main/java/com/kaltura/playkit/player/TrackSelectionHelper.java]: done

**Review by noamtamim — APPROVED** [review]

**Pull request merged**